File: src/main/scala/co.blocke.edi4s/table/TextRenderer.scala
package co.blocke.edi4s
package table

case class TextRenderer(table: Table) extends Renderer:
  private val RESET = "\u001b[0m"

  def render(): String =
    val sb = new StringBuilder
    val widths = normalizedColumnWidths
    val decimalAlignedCellOffsets = findDecimalAlignedCells

    table.title.foreach(t => sb.append(renderTitle(t, widths)))
    sb.append(renderDivider(widths, '='))
    sb.append(renderRow(table.header, widths, decimalAlignedCellOffsets))
    sb.append(renderDivider(widths, '='))

    val rowCount = table.rows.length
    table.rows.zipWithIndex.map{
      case (r: SubHeader,i) =>
        if i > 0 && !isSubHeader(table.rows(i-1)) then
          sb.append(renderDivider(widths))
        sb.append(renderRow(r, widths, decimalAlignedCellOffsets))
        sb.append(renderDivider(widths))
      case (r: Row,i) =>
        sb.append(renderRow(r, widths, decimalAlignedCellOffsets))
        if i == table.rows.length-1 then
          sb.append(renderDivider(widths))
    }
    sb.toString

  private def isSubHeader(r: BodyRow): Boolean =
    r match {
      case _: SubHeader => true
      case _ => false
    }

  private def findDecimalAlignedCells: List[(Int, Int)] =
    val allRows = table.header +: table.rows
    val columnCount = table.columnWidthPct.length

    (0 until columnCount).flatMap { colIdx =>
      var maxRightOfDecimal = 0
      allRows.foreach { row =>
        if colIdx < row.cells.length then
          val cell = row.cells(colIdx)
          val alignment = effectiveAlign(cell, row.align)
          if alignment == Align.DECIMAL then
            val content = cell.content.trim
            val parts = content.split("\\.")
            if parts.length == 2 then
              val rightDigits = parts(1).length
              maxRightOfDecimal = math.max(maxRightOfDecimal, rightDigits)
      }
      if maxRightOfDecimal > 0 then
        Some(colIdx -> maxRightOfDecimal)
      else None
    }.toList

  def normalizedColumnWidths: List[Int] =
    val totalPct = table.columnWidthPct.sum
    table.columnWidthPct.map(p => (p.toDouble / totalPct * table.tableWidth).toInt)

  def effectiveStyle(style: Style): String =
    table.styleMap.getOrElse("text", Map()).getOrElse(style, "")

  def effectiveAlign(cell: Cell, rowAlign: Align): Align =
    cell.align.getOrElse(rowAlign)

  private def renderTitle(title: Title, widths: List[Int]): String =
    val totalWidth = widths.sum + widths.length + 1
    val content = title.cells.head.content
    val pad = (totalWidth - content.length) / 2
    " " * pad + content + "\n"

  private def renderDivider(widths: List[Int], char: Char = '-'): String =
    widths.map(w => "+" + char.toString * w).mkString + "+\n"

  private def padDecimal(content: String, width: Int, rightOffset: Int): String =
    val stripped = content.trim
    val (intPart, fracPart) = stripped.split("\\.") match
      case Array(intP, fracP) => (intP, fracP)
      case Array(intP) => (intP, "")

    val contentWithDot = if fracPart.nonEmpty then s"$intPart.$fracPart" else intPart

    val truncated =
      if contentWithDot.length > width then
        val cutoff = width - 3
        if cutoff <= 0 then "..." else contentWithDot.take(cutoff) + "..."
      else contentWithDot

    val totalLength = truncated.length
    val leftPad = width - math.max(totalLength, rightOffset + intPart.length + (if fracPart.nonEmpty then 1 else 0))
    val rightPad = width - (leftPad + totalLength)

    " " * leftPad + truncated + " " * rightPad

  private def pad(content: String, width: Int, align: Align): String =
    val truncated = if content.length > width then content.take(width - 3) + "..." else content

    align match
      case Align.LEFT =>
        truncated.padTo(width, ' ')
      case Align.CENTER =>
        val padLeft = (width - truncated.length) / 2
        val padRight = width - truncated.length - padLeft
        " " * padLeft + truncated + " " * padRight
      case Align.RIGHT =>
        " " * (width - truncated.length) + truncated
      case Align.DECIMAL =>
        pad(truncated, width, Align.RIGHT) // fallback if no decimal info is known

  private def renderRow(row: TableRow, widths: List[Int], decimalAlignOffsets: List[(Int, Int)]): String =
    val sb = new StringBuilder
    sb.append("|")

    val totalCols = widths.length
    val cellCount = row.cells.length

    var colIndex = 0
    val isCollapsed = cellCount < totalCols

    row.cells.zipWithIndex.foreach { case (cell, i) =>
      val isLastCollapsed = isCollapsed && i == cellCount - 1
      val span = if isLastCollapsed then totalCols - colIndex else math.max(cell.colspan, 1)

      val targetWidth = widths.slice(colIndex, colIndex + span).sum + (span - 1) // account for dividers
      val styleStr = cell.style.map(effectiveStyle).getOrElse(effectiveStyle(row.style))
      val align = effectiveAlign(cell, row.align)
      val content = " " * (cell.indent * 2) + cell.content

      val padStr =
        if align == Align.DECIMAL && decimalAlignOffsets.exists(_._1 == colIndex) then
          val offset = decimalAlignOffsets.find(_._1 == colIndex).get._2
          padDecimal(content, targetWidth, offset)
        else
          pad(content, targetWidth, align)

      sb.append(styleStr).append(padStr).append(RESET).append("|")
      colIndex += span
    }

    sb.append("\n")
    sb.toString

File: src/main/java/co/blocke/edi4j/table/TextRenderer.java
package co.blocke.edi4j.table;

import java.util.*;
import java.util.AbstractMap.SimpleEntry;
import java.util.stream.Collectors;
import java.util.stream.Stream;
import java.util.Objects;

public class TextRenderer implements Renderer {
    private static final String RESET = "\u001b[0m";
    private final Table table;

    public TextRenderer(Table table) {
        this.table = Objects.requireNonNull(table);
    }

    @Override
    public String render() {
        StringBuilder sb = new StringBuilder();
        List<Integer> widths = normalizedColumnWidths();
        // Compute decimal offsets map for quick lookup
        Map<Integer, Integer> offsetMap = findDecimalAlignedCells().stream()
                .collect(Collectors.toMap(SimpleEntry::getKey, SimpleEntry::getValue));

        // render title lines
        for (Title t : table.getTitle()) {
            sb.append(renderTitle(t, widths));
        }
        sb.append(renderDivider(widths, '='));

        // header
        sb.append(renderRow(table.getHeader(), widths, offsetMap));
        sb.append(renderDivider(widths, '='));

        // body rows
        List<BodyRow> rows = table.getRows();
        for (int i = 0; i < rows.size(); i++) {
            BodyRow r = rows.get(i);
            if (r instanceof SubHeader) {
                if (i > 0 && !(rows.get(i - 1) instanceof SubHeader)) {
                    sb.append(renderDivider(widths));
                }
                sb.append(renderRow(r, widths, offsetMap));
                sb.append(renderDivider(widths));
            } else { // Row
                sb.append(renderRow(r, widths, offsetMap));
                if (i == rows.size() - 1) {
                    sb.append(renderDivider(widths));
                }
            }
        }

        return sb.toString();
    }

    private List<SimpleEntry<Integer, Integer>> findDecimalAlignedCells() {
        List<TableRow> all = new ArrayList<>();
        all.add(table.getHeader());
        all.addAll(table.getRows());
        int cols = table.getColumnWidthPct().size();
        List<SimpleEntry<Integer, Integer>> result = new ArrayList<>();

        for (int col = 0; col < cols; col++) {
            int maxRight = 0;
            for (TableRow row : all) {
                if (col < row.getCells().size()) {
                    Cell cell = row.getCells().get(col);
                    Align a = effectiveAlign(cell, row.getAlign());
                    if (a == Align.DECIMAL) {
                        String c = cell.getContent().trim();
                        String[] parts = c.split("\\.");
                        if (parts.length == 2) {
                            maxRight = Math.max(maxRight, parts[1].length());
                        }
                    }
                }
            }
            if (maxRight > 0) {
                result.add(new SimpleEntry<>(col, maxRight));
            }
        }
        return result;
    }

    @Override
    public List<Integer> normalizedColumnWidths() {
        int totalPct = table.getColumnWidthPct().stream().mapToInt(Integer::intValue).sum();
        return table.getColumnWidthPct().stream()
                .map(p -> (int) (p.doubleValue() / totalPct * table.getTableWidth()))
                .collect(Collectors.toList());
    }

    @Override
    public String effectiveStyle(Style style) {
        return table.getStyleMap()
                .getOrDefault("text", Collections.emptyMap())
                .getOrDefault(style, "");
    }

    @Override
    public Align effectiveAlign(Cell cell, Align rowAlign) {
        return cell.getAlign().orElse(rowAlign);
    }

    private String renderTitle(Title title, List<Integer> widths) {
        int total = widths.stream().mapToInt(Integer::intValue).sum() + widths.size() + 1;
        String content = title.getCells().get(0).getContent();
        int pad = (total - content.length()) / 2;
        return " ".repeat(Math.max(0, pad)) + content + "\n";
    }

    private String renderDivider(List<Integer> widths) {
        return renderDivider(widths, '-');
    }

    private String renderDivider(List<Integer> widths, char ch) {
        String line = widths.stream()
                .map(w -> "+" + String.valueOf(ch).repeat(w))
                .collect(Collectors.joining());
        return line + "+\n";
    }

    private String padDecimal(String content, int width, int rightOffset) {
        String s = content.trim();
        String[] parts = s.split("\\.");
        String intPart = parts[0];
        String fracPart = parts.length == 2 ? parts[1] : "";
        String withDot = fracPart.isEmpty() ? intPart : intPart + "." + fracPart;

        String truncated;
        if (withDot.length() > width) {
            int cut = width - 3;
            truncated = cut <= 0 ? "..." : withDot.substring(0, cut) + "...";
        } else {
            truncated = withDot;
        }

        int totalLen = truncated.length();
        int leftPad = width - Math.max(totalLen, rightOffset + intPart.length() + (fracPart.isEmpty() ? 0 : 1));
        int rightPad = width - (leftPad + totalLen);

        return " ".repeat(Math.max(0, leftPad))
                + truncated
                + " ".repeat(Math.max(0, rightPad));
    }

    private String pad(String content, int width, Align align) {
        String t = content.length() > width
                ? content.substring(0, Math.max(0, width - 3)) + "..."
                : content;
        int len = t.length();
        switch (align) {
            case LEFT:
                return t + " ".repeat(Math.max(0, width - len));
            case CENTER: {
                int padL = (width - len) / 2;
                int padR = width - len - padL;
                return " ".repeat(Math.max(0, padL))
                        + t
                        + " ".repeat(Math.max(0, padR));
            }
            case RIGHT:
                return " ".repeat(Math.max(0, width - len)) + t;
            default: // DECIMAL fallback
                return pad(t, width, Align.RIGHT);
        }
    }

    private String renderRow(TableRow row,
                             List<Integer> widths,
                             Map<Integer, Integer> offsetMap) {
        StringBuilder sb = new StringBuilder();
        sb.append("|");

        int totalCols = widths.size();
        int cellCount = row.getCells().size();
        boolean collapsed = cellCount < totalCols;

        int colIdx = 0;
        for (int i = 0; i < cellCount; i++) {
            Cell cell = row.getCells().get(i);
            boolean lastCollapsed = collapsed && i == cellCount - 1;
            int span = lastCollapsed
                    ? totalCols - colIdx
                    : Math.max(cell.getColspan(), 1);

            int target = 0;
            for (int j = colIdx; j < colIdx + span; j++) target += widths.get(j);
            target += (span - 1);

            String styleStr = cell.getStyle()
                    .map(this::effectiveStyle)
                    .orElse(effectiveStyle(row.getStyle()));
            Align align = effectiveAlign(cell, row.getAlign());
            String content = " ".repeat(cell.getIndent() * 2) + cell.getContent();

            Integer offset = offsetMap.get(colIdx);
            String padded;
            if (align == Align.DECIMAL && offset != null) {
                padded = padDecimal(content, target, offset);
            } else {
                padded = pad(content, target, align);
            }

            sb.append(styleStr)
                    .append(padded)
                    .append(RESET)
                    .append("|");
            colIdx += span;
        }
        sb.append("\n");
        return sb.toString();
    }
}